core/templates/dev/head/pages/exploration_editor/simple_editor_tab/ScrollSyncService.js

```diff
@@ -0,0 +1,126 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service that syncs the position of the user in the simple editor
+ with the navigation sidebar.
+ */
+
+oppia.factory('ScrollSyncService', [
+  '$anchorScroll', '$document', '$rootScope', 'QuestionHashService',
+  'SimpleEditorManagerService',
+  function($anchorScroll, $document, $rootScope, QuestionHashService,
+           SimpleEditorManagerService) {
+    var targets = [];
+    var questionList = SimpleEditorManagerService.getQuestionList();
+    var clear = function() {
+      angular.element('.highlighted').removeClass('highlighted');
+    };
+    var collapse = function(hash) {
+      var question;
+      var questionIndex;
+      var questions = questionList.getQuestions();
+      for (var i = 0; i < questions.length; i++) {
+        if (QuestionHashService.getHash(questions[i]) === hash) {
+          questionIndex = i;
+          break;
+        }
+      };
+      if (!angular.isDefined(questionIndex)) {
+        return;
+      };
+      question = questions[questionIndex];
+      $rootScope.$broadcast('SimpleEditorSidebarLinkCollapse', question);
+    };
+    $anchorScroll.yOffset = 80;
+
+    var scrollSyncService = {
+      // Add an element to the targets list, when a user scrolls to an element
+      // in the target list, the sidebar is updated accordingly.
+      addTarget: function(target) {
+        targets.push(target);
+      },
+      // Update the active nav link
+      updateActive: function(hash) {
+        if (!angular.isDefined(hash)) {
+          return;
+        } else {
+          clear();
+          var elm = angular.element('#si-' + hash);
+
+          if (!(hash === 'intro' || hash === 'title')) {
+            elm.addClass('highlighted');
+          };
+
+          // Uncollapse the nav link in case hidden
+          if (elm.is(':hidden')) {
+            collapse(QuestionHashService.getParentQuestionHash(hash));
+          };
+          // TODO: scroll to the nav link in case not displayed
+        }
+      },
+      // Scroll to an element in the editor given it's id.
+      scrollTo: function(hash) {
+        $anchorScroll(hash);
+      }
+    };
+
+    // Onscroll event handler
+    $document.on('scroll', function() {
+      var documentY = $document.scrollTop();
+      var positiveTargets = {};
+      var positiveTargetsDy = [];
+      // Find the closest positive target (the closest element in the target
+      // list to the top of the editor, with a positive distance from it.)
+      for (var i = 0; i < targets.length; i++) {
+        var target = angular.element(targets[i]);
+        var targetOffset = target.offset().top;
+        var dy = targetOffset - documentY;
+        if (dy > 0) {
+          positiveTargets[dy] = target;
+          positiveTargetsDy.push(dy);
+        }
+      };
+      for (var i = 0, minPositive = null; i < positiveTargetsDy.length; i++) {
```

Comment:
Unless I'm missing something, I think all this is a bit too complicated. Just keep track of the minPositiveTarget and its dy; iterate through the list of targets, and update these two values. You can do this in a single for loop; you don't need to enumerate everything first and do another for loop to process those values later.